aptos-move/framework/aptos-stdlib/sources/data_structures/smart_table.move

```diff
@@ -180,6 +187,81 @@ module aptos_std::smart_table {
         res
     }
 
+    /// Get all keys in a smart table.
+    ///
+    /// For a large enough smart table this function will fail due to execution gas limits, and
+    /// `keys_paginated` should be used instead.
+    public fun keys<K: store + copy + drop, V: store + copy>(
+        table_ref: &SmartTable<K, V>
+    ): vector<K> {
+        let (keys, _, _) = keys_paginated(table_ref, 0, 0, ALL_KEYS);
+        keys
+    }
+
+    /// Get keys from a smart table, paginated.
+    ///
+    /// This function can be used to paginate all keys in a large smart table outside of runtime,
+    /// e.g. through chained view function calls. The maximum `num_keys_to_get` before hitting gas
+    /// limits depends on the data types in the smart table.
+    ///
+    /// When starting pagination, pass `starting_bucket_index` = `starting_vector_index` = 0.
+    ///
+    /// The function will then return a vector of keys, an optional bucket index, and an optional
+    /// vector index. The unpacked return indices can then be used as inputs to another pagination
+    /// call, which will return a vector of more keys. This process can be repeated until the
+    /// returned bucket index and vector index value options are both none, which means that
+    /// pagination is complete. For an example, see `test_keys()`.
+    public fun keys_paginated<K: store + copy + drop, V: store + copy>(
+        table_ref: &SmartTable<K, V>,
+        starting_bucket_index: u64,
+        starting_vector_index: u64,
+        num_keys_to_get: u64,
+    ): (
+        vector<K>,
+        Option<u64>,
+        Option<u64>,
+    ) {
+        let keys = vector[];
+        if (num_keys_to_get > 0) {
+            let num_buckets = table_ref.num_buckets;
+            let num_keys_checked = 0;
+            let buckets_ref = &table_ref.buckets;
+            assert!(
+                starting_bucket_index == 0 || starting_bucket_index < num_buckets,
+                EINVALID_BUCKET_INDEX
+            );
+            let bucket_ref = table_with_length::borrow(buckets_ref, starting_bucket_index);
+            let bucket_length = vector::length(bucket_ref);
+            assert!(
+                starting_vector_index == 0 || starting_vector_index < bucket_length,
+                EINVALID_VECTOR_INDEX
+            );
+            for (bucket_index in starting_bucket_index..num_buckets) {
+                bucket_ref = table_with_length::borrow(buckets_ref, bucket_index);
+                bucket_length = vector::length(bucket_ref);
+                for (vector_index in starting_vector_index..bucket_length) {
+                    vector::push_back(&mut keys, vector::borrow(bucket_ref, vector_index).key);
+                    num_keys_checked = num_keys_checked + 1;
```

Comment:
why not if `--num_keys_to_get == 0`? I don't think we need num_keys_checked if `num_keys_to_get >= 1`

Reply:
Addressed in 063ebdd906